mobile/src/main/java/ch/epfl/sdp/mobile/ui/game/ar/ChessScene.kt

```diff
@@ -1,74 +1,123 @@
 ) {
-  val boardNode: ArModelNode
+  val boardNode: ArModelNode = ArModelNode(placementMode = PlacementMode.PLANE_HORIZONTAL)
+
+  private var currentPosition: Map<Position, Piece> = emptyMap()
+  private var currentPieces: MutableMap<Piece, ModelNode> = mutableMapOf()
 
   private var boardHeight: Float = 0f
   private var boardHalfSize: Float = 0f
 
+  private var isLoaded = false
+
   init {
-    boardNode =
-        ArModelNode(placementMode = PlacementMode.PLANE_HORIZONTAL).apply {
-          // Load the board
-          loadModelAsync(
-              context = context,
-              glbFileLocation = ChessModels.Board,
-              coroutineScope = lifecycleScope,
-              autoScale = true,
-          ) { renderableInstance ->
-            // Once loaded compute the board size
-            val filamentAsset = renderableInstance.filamentAsset ?: return@loadModelAsync
-            val boardBoundingBox = filamentAsset.boundingBox
-
-            // Get height (on y axe) of the board
-            // Double the value to get the total height of the box
-            boardHeight = 2 * boardBoundingBox.halfExtent[1]
-            boardHalfSize = boardBoundingBox.halfExtent[0]
-
-            // Initialize all pieces
-            for ((position, piece) in board) {
-              val model = loadPieceModel(piece, position)
-              // Add the new model node to the board
-              addChild(model)
-            }
-          }
-        }
+    scope.launch {
+      // Load Board
+      val renderableInstance = loadBoard() ?: return@launch
+
+      // Once loaded compute the board size
+      val filamentAsset = renderableInstance.filamentAsset ?: return@launch
+      val boardBoundingBox = filamentAsset.boundingBox
+
+      // Get height (on y axe) of the board
+      // Double the value to get the total height of the box
+      boardHeight = 2 * boardBoundingBox.halfExtent[1]
+      boardHalfSize = boardBoundingBox.halfExtent[0]
+
+      // Load pieces
+      loadPieces(startingBoard)
+      isLoaded = true
+
+      // Move the pieces to the correct position
+      moveAllPiece()
+    }
+  }
+
+  /**
+   * Use by [ArChessBoardScreen] to update the pieces' position on the board
+   *
+   * @param pieces The new piece position
+   */
+  fun update(pieces: Map<Position, Piece>) {
+    savePosition(pieces)
+    moveAllPiece()
+  }
+
+  /**
+   * Save the newest pieces position
+   *
+   * @param pieces The new pieces position to save
+   */
+  private fun savePosition(pieces: Map<Position, Piece>) {
+    currentPosition = pieces
+  }
+
+  /** Move all pieces */
+  private fun moveAllPiece() {
+    if (isLoaded) {
+      for ((position, piece) in currentPosition) {
+        move(piece, position)
+      }
+    }
+  }
+
+  /** Move the given [piece] at a specific [position]] */
+  private fun move(piece: Piece, position: Position) {
+    val model = currentPieces[piece] ?: return
+    model.position = toArPosition(position)
+  }
```

Comment:
Question : is `isLoaded` actually needed ? If the piece models are not loaded yet, they won't be present in `currentPieces` and `move` will no-op (since `currrentPieces[piece] ?: return` will return).